packages/nextjs/src/utils/handlers.ts

```diff
@@ -24,73 +30,113 @@ export const withSentry = (handler: NextApiHandler): WrappedNextApiHandler => {
+          currentScope.setSpan(transaction);
+
+          // save a link to the scope on the response, so that even if there's an error (landing us outside of
+          // the domain), we can still finish the transaction and attach the correct data to it
+          (res as ScopedResponse).__sentryScope = currentScope;
         }
+      }
 
+      try {
         return await handler(req, res); // Call original handler
       } catch (e) {
-        withScope(scope => {
-          scope.addEventProcessor(event => {
+        if (currentScope) {
+          currentScope.addEventProcessor(event => {
             addExceptionMechanism(event, {
               handled: false,
             });
-            return parseRequest(event, req);
+            return event;
           });
           captureException(e);
-        });
-        throw e;
-      } finally {
-        const transaction = getActiveTransaction();
-        if (transaction) {
-          transaction.setHttpStatus(res.statusCode);
-
-          transaction.finish();
-        }
-        try {
-          await flush(2000);
-        } catch (e) {
-          // no-empty
         }
+        throw e;
       }
     });
 
     return await boundHandler();
   };
 };
+
+type ResponseEndMethod = ScopedResponse['end'];
+type WrappedResponseEndMethod = ResponseEndMethod;
+
+function wrapEndMethod(origEnd: ResponseEndMethod): WrappedResponseEndMethod {
+  return async function newEnd(this: ScopedResponse, ...args: unknown[]) {
+    // if the handler errored, it will have popped us out of the domain, so push the domain's scope onto the stack
+    // just in case (if we *are* still in the domain, this will replace the current scope with a clone of itself,
+    // which is effectively a no-op as long as we remember to pop it off when we're done)
+    const currentHub = getCurrentHub();
+    currentHub.pushScope(this.__sentryScope);
+
+    const transaction = getActiveTransaction();
+
+    if (transaction) {
+      transaction.setHttpStatus(this.statusCode);
+
+      // Push `transaction.finish` to the next event loop so open spans have a better chance of finishing before the
+      // transaction closes, and make sure to wait until that's done before flushing events
+      const transactionFinished: Promise<void> = new Promise(resolve => {
+        setImmediate(() => {
+          transaction.finish();
+          resolve();
+        });
+      });
+      await transactionFinished;
+    }
+
+    // flush the event queue to ensure that events get sent to Sentry before the response is finished and the lambda
+    // ends
+    try {
+      logger.log('Flushing events...');
+      await flush(2000);
+      logger.log('Done flushing events');
+    } catch (e) {
+      logger.log(`Error while flushing events:\n${e}`);
+    }
+
+    // now that our work is done, we can pop off the scope and allow the response to end
+    if (currentHub.getScope()?.getParent() === this.__sentryScope) {
+      currentHub.popScope();
```

Comment:
The `pushScope` and `popScope` in this block is unsafe business.

It temporarily affects global state by changing the perceived current scope, and if we ever got a report per the warning message below, there'd be not much for us to do.

I'd be more comfortable with code that only deals with spans and leave scopes/hubs alone for now.